base/strings/styled.jl

```diff
@@ -0,0 +1,363 @@
+end
+
+"""
+    StyledChar{S <: AbstractChar} <: AbstractChar
+
+A Char annotated by properties (often styling information).
+
+More specifically, this is a thin wrapper around any other [`AbstractChar`](@ref),
+which adds arbitrary named annotations to the wrapped character.
+
+## Constructors
+
+```julia
+StyledChar(s::S) -> StyledChar{S}
+StyledChar(s::S, props::Pair{Symbol, <:Any}...)
+StyledChar(s::S, properties::Vector{Pair{Symbol, <:Any}})
+```
+"""
+struct StyledChar{C <: AbstractChar} <: AbstractChar
+    char::C
+    properties::Vector{Pair{Symbol, Any}}
+end
+
+## Constructors ##
+
+StyledString(s::AbstractString, prop::Pair{Symbol, <:Any}, props::Pair{Symbol, <:Any}...) =
+    StyledString(s, firstindex(s):lastindex(s), prop, props...)
+
+StyledString(s::AbstractString, region::UnitRange{Int}, props::Pair{Symbol, <:Any}...) =
+    StyledString(s, [(region, Pair{Symbol, Any}(first(p), last(p)))
+                      for p in props])
+
+StyledString(s::AbstractString, props::Vector{<:Pair{Symbol, <:Any}}) =
+    StyledString(s, [(firstindex(s):lastindex(s), p) for p in props])
+
+# Constructors called with overly-specialised arguments
+
+StyledString(s::AbstractString, props::Vector{<:Tuple{UnitRange{Int}, <:Pair{Symbol, <:Any}}}) =
+    StyledString(s, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}(props))
+
+StyledChar(c::AbstractChar, prop::Pair{Symbol, <:Any}, props::Pair{Symbol, <:Any}...) =
+    StyledChar(c, Vector{Pair{Symbol, Any}}(vcat(prop, props...)))
+
+# Constructors to avoid recursive wrapping
+
+StyledString(s::StyledString, props::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}) =
+    StyledString(s.string, vcat(s.properties, props))
+
+StyledChar(c::StyledChar, props::Vector{Pair{Symbol, Any}}) =
+    StyledChar(c.char, vcat(s.properties, props))
+
+# To avoid pointless overhead
+String(s::StyledString{String}) = s.string
+
+## Conversion/promotion ##
+
+convert(::Type{StyledString}, s::StyledString) = s
+convert(::Type{StyledString{S}}, s::S) where {S <: AbstractString} =
+    StyledString(s, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+convert(::Type{StyledString}, s::S) where {S <: AbstractString} =
+    convert(StyledString{S}, s)
+StyledString(s::S) where {S <: AbstractString} = convert(StyledString{S}, s)
+
+convert(::Type{StyledChar}, c::StyledChar) = c
+convert(::Type{StyledChar{C}}, c::C) where { C <: AbstractChar } =
+    StyledChar{C}(c, Vector{Pair{Symbol, Any}}())
+convert(::Type{StyledChar}, c::C) where { C <: AbstractChar } =
+    convert(StyledChar{C}, c)
+
+StyledChar(c::AbstractChar) = convert(StyledChar, c)
+StyledChar(c::UInt32) = convert(StyledChar, Char(c))
+StyledChar{C}(c::UInt32) where {C <: AbstractChar} = convert(StyledChar, C(c))
+
+promote_rule(::Type{<:StyledString}, ::Type{<:AbstractString}) = StyledString
+
+## AbstractString interface ##
+
+ncodeunits(s::StyledString) = ncodeunits(s.string)
+codeunits(s::StyledString) = codeunits(s.string)
```

Comment:
Are all these needed? Doesn't the `AbstractString` interface provide some of these?

Reply:
There isn't a default implementation, without out these you get:

```julia

ERROR: MethodError: no method matching ncodeunits(::StyledString{String})

Closest candidates are:
  ncodeunits(::SubString)
   @ Base strings/substring.jl:81
  ncodeunits(::StyledChar)
   @ Base ~/Desktop/TEC/Projects/available/julia/base/strings/styled.jl:121
  ncodeunits(::SubstitutionString)
   @ Base regex.jl:568
  ...
```